// pupsik/core/process/physic/unit/units/Length.js
const Unit = require('../core/Unit');

let Length = function ( ) { };
let NewUnit = Length;

NewUnit.prototype = Unit;
Unit.call( NewUnit );


NewUnit.push({
    name: 'meter',
    symbol: 'm',
    defaultPower: 1,
    power: 1,
    conversionLvl: 1000,
    toMain: 1,
    fromMain: 1,

    isPrefixSupport: true,
});

NewUnit.push({
    name: 'area',
    symbol: 'm',
    defaultPower: 2,
    power: 2,
    conversionLvl: 1000,
    toMain: 1,
    fromMain: 1,

    isPrefixSupport: true,
});

NewUnit.push({
    name: 'volume',
    symbol: 'm',
    defaultPower: 3,
    power: 3,
    conversionLvl: 1000,
    toMain: 1,
    fromMain: 1,

    isPrefixSupport: true,
});

NewUnit.push({
    name: 'nail',
    symbol: 'nail',
    defaultPower: 1,
    power: 1,
    conversionLvl: 0,
    toMain: 1/17.486,
    fromMain: 17.486,
    conversionLvl: 0,
});

NewUnit.push({
    name: 'hectare',
    symbol: 'ha',
    defaultPower: 2,
    power: 1,
    conversionLvl: 0,
    toMain: 100,
    fromMain: 1/100,
});

NewUnit.update();

module.exports = Length;
